File: edu/api/teachers.py
#  Nikulin Vasily © 2021
from flask_login import current_user

from data import db_session
from data.roles import Role
from data.users import User
from edu.api import api, socket
from tools.tools import send_response, fillJson, roles_allowed


@socket.on('getTeachers')
@api.route('/api/teachers', methods=['GET'])
def getTeachers():
    event_name = 'getTeachers'

    db_sess = db_session.create_session()

    school_id = current_user.school_id
    teachers = list(filter(lambda user: user.has_role('teacher') or user.has_role('head_teacher'),
                           db_sess.query(User).filter(
                               User.school_id == school_id
                           ).all()))

    roles = list(map(lambda x: x[0], db_sess.query(Role.name).all()))
    teachers_data = []
    for teacher in teachers:
        teacher_roles = []
        for role in roles:
            if teacher.has_role(role):
                teacher_roles.append(role)
        teacher_dict = teacher.to_dict(only=('id', 'surname', 'name', 'patronymic', 'date_of_birth',
                                             'email', 'school_id', 'about', 'game_session_id'))
        teacher_dict['roles'] = teacher_roles
        teachers_data.append(teacher_dict)

    return send_response(
        event_name,
        {
            'message': 'Success',
            'users': teachers_data
        }
    )


@socket.on('editTeacher')
@api.route('/api/teachers', methods=['PUT'])
@roles_allowed('head_teacher', 'director')
def editTeacher(json=None):
    event_name = 'editTeacher'

    db_sess = db_session.create_session()

    keys = ['surname', 'name', 'patronymic', 'roles']
    fillJson(json, keys)

    teacher = db_sess.query(User).filter(
        User.school_id == current_user.school_id,
        User.surname == json['surname'],
        User.name == json['name'],
    ).first()

    teacher.patronymic = json['patronymic']
    teacher.clear_roles(['student', 'teacher', 'head_teacher'])
    teacher.add_roles(json['roles'])

    db_sess.merge(teacher)
    db_sess.commit()

    return send_response(
        event_name,
        {
            'message': 'Success',
        }
    )


@socket.on('deleteTeacher')
@api.route('/api/teachers', methods=['DELETE'])
@roles_allowed('head_teacher', 'director')
def deleteTeacher(json=None):
    event_name = 'deleteTeacher'

    db_sess = db_session.create_session()

    keys = ['surname', 'name', 'patronymic', 'roles']
    fillJson(json, keys)

    teacher = db_sess.query(User).filter(
        User.school_id == current_user.school_id,
        User.surname == json['surname'],
        User.name == json['name'],
    ).first()

    teacher.patronymic = json['patronymic']
    teacher.clear_roles(['student', 'teacher', 'head_teacher'])

    db_sess.merge(teacher)
    db_sess.commit()

    return send_response(
        event_name,
        {
            'message': 'Success',
        }
    )

File: edu/static/js/api/teachers.js
/*
 * Nikulin Vasily © 2021
 */

teachers = Object()
teachers.get = function (fn = null) {
    if (fn)
        teachers.getFn = fn
    else
        teachers.getFn = null
    socket.emit('getTeachers')
}

socket.on('getTeachers', function (data) {
    teachers.getJson = data
    if (teachers.getFn)
        teachers.getFn()
})

teachers.post = function (surname = null,
                          name = null,
                          patronymic = null,
                          roles = ['user', 'teacher'],
                          fn = null) {
    if (fn)
        teachers.postFn = fn
    else
        teachers.postFn = null
    let email = translit(surname) + translit(name)[0].toUpperCase() + '@area-146.tk'
    let password = translit(surname) + translit(name)[0].toUpperCase() + '_password'
    socket.emit('createUser', {
        'email': email,
        'password': password,
        'surname': surname,
        'name': name,
        'patronymic': patronymic,
        'roles': roles,
        'school_id': 'current'
    })
}

socket.on('createUser', function (data) {
    teachers.postJson = data
    if (teachers.postFn)
        teachers.postFn(data)
})

teachers.put = function (surname = null,
                         name = null,
                         patronymic = null,
                         roles = ['user', 'teacher'],
                         fn = null) {
    if (fn)
        teachers.putFn = fn
    else
        teachers.putFn = null

    socket.emit('editTeacher', {
        'surname': surname,
        'name': name,
        'patronymic': patronymic,
        'roles': roles
    })
}

socket.on('editTeacher', function (data) {
    teachers.putJson = data
    if (teachers.putFn)
        teachers.putFn(data)
})

teachers.delete = function (surname = null,
                            name = null,
                            patronymic = null,
                            roles = null,
                            fn = null) {
    if (fn)
        teachers.deleteFn = fn
    else
        teachers.deleteFn = null

    socket.emit('deleteTeacher', {
        'surname': surname,
        'name': name,
        'patronymic': patronymic,
        'roles': roles
    })
}

socket.on('deleteTeacher', function (data) {
    teachers.deleteJson = data
    if (teachers.deleteFn)
        teachers.deleteFn(data)
})

function translit(word) {
    let answer = '';
    const converter = {
        'а': 'a', 'б': 'b', 'в': 'v', 'г': 'g', 'д': 'd',
        'е': 'e', 'ё': 'e', 'ж': 'zh', 'з': 'z', 'и': 'i',
        'й': 'y', 'к': 'k', 'л': 'l', 'м': 'm', 'н': 'n',
        'о': 'o', 'п': 'p', 'р': 'r', 'с': 's', 'т': 't',
        'у': 'u', 'ф': 'f', 'х': 'h', 'ц': 'c', 'ч': 'ch',
        'ш': 'sh', 'щ': 'sch', 'ь': '', 'ы': 'y', 'ъ': '',
        'э': 'e', 'ю': 'yu', 'я': 'ya',

        'А': 'A', 'Б': 'B', 'В': 'V', 'Г': 'G', 'Д': 'D',
        'Е': 'E', 'Ё': 'E', 'Ж': 'Zh', 'З': 'Z', 'И': 'I',
        'Й': 'Y', 'К': 'K', 'Л': 'L', 'М': 'M', 'Н': 'N',
        'О': 'O', 'П': 'P', 'Р': 'R', 'С': 'S', 'Т': 'T',
        'У': 'U', 'Ф': 'F', 'Х': 'H', 'Ц': 'C', 'Ч': 'Ch',
        'Ш': 'Sh', 'Щ': 'Sch', 'Ь': '', 'Ы': 'Y', 'Ъ': '',
        'Э': 'E', 'Ю': 'Yu', 'Я': 'Ya'
    };

    for (let i = 0; i < word.length; ++i) {
        if (converter[word[i]] === undefined) {
            answer += word[i];
        } else {
            answer += converter[word[i]];
        }
    }

    return answer;
}